src/log/File.cc

```diff
@@ -112,21 +112,35 @@ void
 logfilePrintf(Logfile * lf, const char *fmt,...)
 {
     va_list args;
-    char buf[8192];
-    int s;
-
     va_start(args, fmt);
 
-    s = vsnprintf(buf, 8192, fmt, args);
-
-    if (s > 8192) {
-        s = 8192;
+    // bufsize will be large enough for most log lines
+    int bufsize = 16 * 1024;
+    char* buf = NULL;
+
+    while (true) {
+        buf = new char[bufsize];
```

Comment:
Buffers allocated on stacks are the #1 cause of vulnerabilities in software.  It is much safer to allocate memory in the heap.  Also, 16K is an excessive amount of memory to allocate on the stack (the former amount was 8K, which is also excessive for the stack).  Allocating a static buffer would be fine, but then the function is no longer thread-safe, unless the memory can be allocated with the C++11 thread_local directive.

Reply:
AFAIK, this function does not need to be thread safe. In fact, virtually no Squid function needs to be thread safe except a very small set of specialized functions dedicated to AUFS code (src/DiskIO/DiskThreads/). Outside of threaded AUFS code, virtually any "I will do X because of thread safety" argument is invalid. This logging code is outside of threaded AUFS code.

And if you convert this code to use SBuf, then we do not need to argue whether allocating X KB on stack is "safe" -- SBuf does not use stack for its storage.